### PR TITLE
Support for ECDSA signatures on curve secp256r1

### LINKED ISSUE #79: ECDSA support
With the simplified cryptographic scheme (#73) of Biscuit 2.0 (#72), it becomes possible to support other cryptographic algorithms.

I would like to make a suggestion to support ECDSA in conjunction with the NIST P-256 elliptic curve. Full disclosure I am a maintainer of the `ecdsa` and `p256` Rust crates.

The main reason for supporting ECDSA/P-256 is compatibility with hardware tokens, smartcards, and SEPs. Notably the following devices support it, and don't support Ed25519:

- Yubikey 4/5 devices
- iOS/macOS Keychain + SEP

The use case would be registering the ECDSA public key associated with a hardware-backed private key and allowing a user to sign Biscuits with that hardware-backed private key.

ECDSA has a somewhat notorious past of sharp edges and implementation mistakes. The two biggest are:

- Dependence on a random scalar `k` which if selected in a biased manner leaks the private key. This is mitigated through using a pseudorandom scalar selected by rejection sampling as standardized in RFC6979, a method not too far off from how Ed25519 works
- Incomplete addition formulas were often used to implement the NIST curves, e.g. Jacobian formulas. I'll note that the `p256` crate at least implements [complete addition formulas for prime order elliptic curves](https://eprint.iacr.org/2015/1060.pdf). Other implementations may vary, although this is generally an issue impacting D-H/encryption rather than signatures

### PR DESCRIPTION
**Pull request by Geal**
Fix #79

To allow Biscuit tokens to be signed by HSM or other key managers, we introduce a way to use ECDSA to sign blocks

TODO:
- [ ] samples
- [ ] https://github.com/biscuit-auth/biscuit-rust/pull/108
- [ ] Go implementation
- [ ] Java implementation

### CODE REVIEW

File: SPECIFICATIONS.md [hunk 1]
```diff
@@ -683,16 +684,58 @@ is a Biscuit token, that base 64 string should be prefixed with `biscuit:`.
...
+##### ECDSA
+
+Biscuit supports the ECDSA algorithm over the secp256r1 curve as defined in
+[SEC2v1](https://www.secg.org/sec1-v2.pdf), using the SHA-256 hashing
+algorithm. It is recommended to use a deterministic signature algorithm
+version like the one defined in [RFC 6979](https://datatracker.ietf.org/doc/html/rfc6979).
+
+
+The protobuf encoding is defined as follows:
+- `key` field of the `Publickey` message: [SEC1 format, defined in section 2.3.3](https://www.secg.org/sec1-v2.pdf). Allowed prefixes: `02`, `03`, `04`
```
Comment (tarcieri) [inline SPECIFICATIONS.md:725]: Do you really need to support uncompressed public keys? They're just a workaround for long-expired point compression patents.

Comment (Geal) [inline SPECIFICATIONS.md:725]:
```suggestion
- `key` field of the `Publickey` message: [compressed SEC1 format, defined in section 2.3.3](https://www.secg.org/sec1-v2.pdf). Allowed prefixes: `02`, `03`
```

Comment (Geal) [reply SPECIFICATIONS.md]: fixed in [bd9901d](https://github.com/biscuit-auth/biscuit/pull/165/commits/bd9901d70a72b3f6c1163114f1eea996ccc289b0)

File: SPECIFICATIONS.md [hunk 2]
```diff
@@ -688,16 +689,58 @@ is a Biscuit token, that base 64 string should be prefixed with `biscuit:`.
...
+Biscuit tokens are based on public key cryptography, with a chain of
 signatures. Each block contains the serialized Datalog, the next public key,
 and the signature by the previous key. The token also contains the private key
 corresponding to the last public key, to sign a new block and attenuate the
 token, or a signature of the last block by the last private key, to seal the
 token.
 
+#### Algorithms
+
+Biscuit supports multiple signature algorithms for it blocks, that can change
```
Comment (Geal) [inline SPECIFICATIONS.md:701]:
```suggestion
Biscuit supports multiple signature algorithms for its blocks, that can change
```

**Comment by divarvel** [comment]
`SPECIFICATIONS.md` looks good, you’ll need to update samples.

**Comment by Geal** [comment]
ok, I think this is in a good shape now. The [ref implementation](https://github.com/biscuit-auth/biscuit-rust/pull/108) is ready, and it is much simpler now wit the refactorings due to the payoad changes

**Pull request merged**